File: windowsforms/matching-game/net5-windows/cs/MatchingGame.Logic/Game.cs
using System;
using System.Collections.Generic;

namespace MatchingGame.Logic
{
    public sealed class Game
    {
        private readonly char[] _cards;
        private readonly bool[] _openCards;
        private readonly List<int> _turnCards = new List<int>(2);
        private int _turns;

        private Game(int width, int height, char[] board)
        {
            Width = width;
            Height = height;
            _cards = board;
            _openCards = new bool[_cards.Length];
        }

        public static Game Create()
        {
            var numberOfPairs = 8;
            var width = numberOfPairs / 2;
            var height = width;

            var board = new char[width * height];
            var availableCards = new List<char>()
            {
                '!', '!', 'N', 'N', ',', ',', 'k', 'k',
                'b', 'b', 'v', 'v', 'w', 'w', 'z', 'z'
            };

            var random = new Random();

            for (var i = 0; i < numberOfPairs * 2; i++)
            {
                var cardIndex = random.Next(availableCards.Count - 1);
                var card = availableCards[cardIndex];
                board[i] = card;
                availableCards.RemoveAt(cardIndex);
            }

            return new Game(width, height, board);
        }

        public int Turns => _turns;

        public int Width { get; }

        public int Height { get; }

        public int RemainingCardsInTurn => 2 - _turnCards.Count;

        private int GetIndex(int w, int h)
        {
            return h * Width + w;
        }

        public int OpenCard(int w, int h)
        {
            if (_turnCards.Count > 1)
                throw new InvalidOperationException("Cannot open more than two cards");

            var cardIndex = GetIndex(w, h);
            var cardValue = _cards[cardIndex];
            _turnCards.Add(cardIndex);
            _openCards[cardIndex] = true;

            if (_turnCards.Count == 2)
                _turns++;

            return cardValue;
        }

        public void CloseCards()
        {
            foreach (var cardIndex in _turnCards)
                _openCards[cardIndex] = false;
        }

        public char GetCard(int w, int h)
        {
            var cardIndex = GetIndex(w, h);
            return _cards[cardIndex];
        }

        public bool IsOpen(int w, int h)
        {
            var v = GetIndex(w, h);
            return _openCards[v];
        }

        public bool CompleteTurn()
        {
            if (_turnCards.Count != 2)
                return false;

            var firstCardIndex = _turnCards[0];
            var secondCardIndex = _turnCards[1];
            var isMatch = _cards[firstCardIndex] == _cards[secondCardIndex];
            if (!isMatch)
                CloseCards();

            _turnCards.Clear();
            return isMatch;
        }

        public bool IsComplete()
        {
            foreach (var isVisible in _openCards)
                if (!isVisible)
                    return false;

            return true;
        }
    }
}

File: windowsforms/matching-game/net5-windows/vb/MatchingGame.Logic/Game.vb
Imports System.Collections.Generic

Public NotInheritable Class Game

    Private ReadOnly _cards() As Char
    Private ReadOnly _openCards() As Boolean
    Private ReadOnly _turnCards As New List(Of Integer)(2)

    Private _turns As Integer

    Private Sub New(width As Integer, height As Integer, board As Char())
        Me.Width = width
        Me.Height = height
        _cards = board
        ReDim _openCards(_cards.Length - 1)
    End Sub

    Public Shared Function Create() As Game
        Dim numberOfPairs = 8
        Dim width = numberOfPairs / 2
        Dim height = width

        Dim board((width * height) - 1) As Char
        Dim availableCards = New List(Of Char) From
        {
            "!"c, "!"c, "N"c, "N"c, ","c, ","c, "k"c, "k"c,
            "b"c, "b"c, "v"c, "v"c, "w"c, "w"c, "z"c, "z"c
        }

        Dim Random As New Random()

        For i = 0 To numberOfPairs * 2 - 1
            Dim cardIndex = random.Next(availableCards.Count - 1)
            Dim card = availableCards(cardIndex)
            board(i) = card
            availableCards.RemoveAt(cardIndex)
        Next

        Return New Game(width, height, board)
    End Function

    Public ReadOnly Property Turns As Integer
        Get
            Return _turns
        End Get
    End Property

    Public ReadOnly Property Width
    Public ReadOnly Property Height
    Public ReadOnly Property RemainingCardsInTurn
        Get
            Return 2 - _turnCards.Count
        End Get
    End Property

    Private Function GetIndex(w As Integer, h As Integer) As Integer
        Return h * Width + w
    End Function

    Public Function OpenCard(w As Integer, h As Integer) As Integer
        If _turnCards.Count > 1 Then
            Throw New InvalidOperationException("Cannot open more than two cards.")
        End If

        Dim cardIndex = GetIndex(w, h)
        Dim cardValue = _cards(cardIndex)

        _turnCards.Add(cardIndex)
        _openCards(cardIndex) = True

        If _turnCards.Count = 2 Then
            _turns += 1
        End If

        Return AscW(cardValue)
    End Function

    Public Sub CloseCards()
        For Each cardIndex In _turnCards
            _openCards(cardIndex) = False
        Next
    End Sub

    Public Function GetCard(w As Integer, h As Integer) As Char
        Dim cardIndex = GetIndex(w, h)
        Return _cards(cardIndex)
    End Function

    Public Function IsOpen(w As Integer, h As Integer) As Boolean
        Dim cardIndex = GetIndex(w, h)
        Return _openCards(cardIndex)
    End Function

    Public Function CompleteTurn() As Boolean
        If _turnCards.Count <> 2 Then
            Return False
        End If

        Dim firstCardIndex = _turnCards(0)
        Dim secondCardIndex = _turnCards(1)
        Dim isMatch = _cards(firstCardIndex) = _cards(secondCardIndex)

        If Not isMatch Then
            CloseCards()
        End If

        _turnCards.Clear()

        Return isMatch
    End Function

    Public Function IsComplete() As Boolean
        For Each isVisible In _openCards
            If Not isVisible Then
                Return False
            End If
        Next

        Return True
    End Function
End Class
